Android 11 Support

whew-inc opened the pull request:
I've deleted the whole package name service thing and used a simpler way to retrieve package names and labels. Both the package and app name can be logged this way. This fixes the original NullPointerException spam in log.

As for Android 11 support: this now hooks into `ApplicationPackageManager.getPackageInfo` instead of the old method.

I've also upped Gradle version and target SDK.

Tested on Android 11 AOSP ROM.
~~Tested on Android 10 Samsung ROM.~~

whew-inc committed:
Fix NullPointerException

whew-inc commented:
Forgot to add that this should fix issue #36.

SynapseScribe commented:
Can we have a new release with the fix please? Edxposed logs are still spammed on android Q

Joecowboy commented:
**_I am running a OnePlus 7 Pro 5G (Sprint) latest firmware De-Googled.  I uninstalled all Google apps and framework._**  

You can do the same thing Running the OnePlus 7 Pro 5G (Sprint) running the EU 5G firmware.  You might want to copy the Modem image from the Sprint's latest firmware using TWRP and then write it back to the EU do to the latest builds (as of v10.0.9) modem is flaky and will not stay connected.

Backup Modem.img to both A and B partitions update using terminal:
dd if=/dev/block/bootdevice/by-name/modem_a of=/sdcard/Download/modem.img 
dd if=/dev/block/bootdevice/by-name/modem_b of=/sdcard/Download/modem.img 

Write Modem.img to both A and B partitions update using terminal:
dd if=/sdcard/Download/modem.img of=/dev/block/bootdevice/by-name/modem_a
dd if=/sdcard/Download/modem.img of=/dev/block/bootdevice/by-name/modem_b

I can confirm **_whew-inc_** will work with MicroG latest build Magisk revived module.  I am using it with Magisk 21.3 and latest build of EdXposed YAHAFA Canary. Just install Ruri modules via Magisk Manager.  Just enabled the FakesGApps module in EdXposed Manager after you install it.  If you run into any issues with it not wanting to check the boxes in MicroG Self-Check might need to fix the permissions XML file after you install it.  Also, with OnePlus 7 Pro and Magisk 21.x version you might run into an issue of not being able to use a root explorer or terminal exploring root hangs you might need to install the latest version of  Disable_Dm-Verity_ForceEncryption.zip follow the steps on XDA forum exactly.

================================================================================
Update the following file:
/data/system/users/0/runtime-permissions.xml

Find this section and update it:
     Change 4300 to 3320 and this should fix issue.  Reboot and you should now be able to check the boxes.  

```
`<pkg` name="com.google.android.gms">
    <item name="android.permission.ACCESS_FINE_LOCATION" granted="true" flags="300" />
    <item name="android.permission.FAKE_PACKAGE_SIGNATURE" granted="true" flags="300" />
    <item name="android.permission.READ_EXTERNAL_STORAGE" granted="true" flags="4300" />
    <item name="android.permission.ACCESS_COARSE_LOCATION" granted="true" flags="300" />
    <item name="android.permission.READ_PHONE_STATE" granted="true" flags="300" />
    <item name="android.permission.GET_ACCOUNTS" granted="true" flags="300" />
    <item name="android.permission.WRITE_EXTERNAL_STORAGE" granted="true" flags="4300" />
    <item name="android.permission.ACCESS_BACKGROUND_LOCATION" granted="false" flags="4300" />
  </pkg>
```

==============================================================================

I have attached compiled version (extract it from the zip):

[FakeGapps.zip](https://github.com/thermatk/FakeGApps/files/5825451/FakeGapps.zip)

downtoground commented:
I have no such permission file in LOS under /data/system/users/0/runtime-permissions.xml

Would you explain how to setup fakegapps in my OS?

tenpura-shrimp commented:
This is a good change. It wasn't working for me on 11 though.  I got it working by modifying the hook to hook ApplicationPackageManger.getPackageInfo

dantecpu commented:
As current status, documentation about proper steps to make it work on Android 11 require improvements, simple compilation and install isn't enough on a current R build for klte.

whew-inc committed:
Initial android 11 support

whew-inc commented:
> This is a good change. It wasn't working for me on 11 though. I got it working by modifying the hook to hook ApplicationPackageManger.getPackageInfo

This seems to work for me too on 11. However I think this won't work on older releases so likely gonna have to add a version check.

I built an apk for testing: https://github.com/whew-inc/FakeGApps/releases/tag/android11_initial

Can someone test this with android < 11?

(If you would rather build it yourself: https://github.com/whew-inc/FakeGApps/tree/android11_initial)

chaoscalm commented:
I self-compiled this for testing on armv7 , device klte LineageOS 18.1 and confirm it works good, I can test on one of my tablets which runs android 9.0 and see if it works same.

chaoscalm commented:
@whew-inc Spoofing works on another tablet running Android 9.0 with https://github.com/whew-inc/FakeGApps/releases/tag/android11_initial for me, needs more tests about this.

whew-inc commented:
Youtube Vanced (official one too?) does not seem to trigger the hook. "YouTube won't run without Google Play services" yadayada

Signal does not work too.

Play services reports services are installed.

My country's corona app works properly.


`ApplicationPackageManager.getPackageInfo` is probably not a good method to hook. I will look into what changed in R.

chaoscalm commented:
About YouTube vanced: https://vancedmanager.com/

there is a known issue with versions > 16 in order to make microg integration work https://github.com/microg/GmsCore/issues/1373 

I personally use an old v15 for the channels with subscriptions, and newpipe for anything else.

Another degoogled signal alternative: https://langis.cloudfrancois.fr/#download-en

Signal works for me, so I guess we are using different microg implementations: https://github.com/FriendlyNeighborhoodShane/MinMicroG-abuse-CI This is what I use.

I used the corona app for my country too, but I ended stopping using it because of the many bugs.

**Edit:**

For ethics reason I'll use the degoogled version above, but I just tested the Google Play one downloaded from Aurora Store, and works fine at first sight.

whew-inc commented:
@dantecpu Thanks for the info. Older Vanced version didn't work at first but after a few reboots and such it magically works now.

Signal is now the only one acting up, though it seems it switched to using the notification trick to push new message notifications. I can't be bothered to backup my messages and install something else right now.

I've also checked the source code a bit and it looks like `ApplicationPackageManager.getPackageInfo` also existed in older Android versions, so I expect it to work on older versions. Though I can't tell why the original hook doesn't work anymore. I don't feel like digging through the source code that much. The only reasons I can think of right now:

Android doesn't use the method anymore
The old method is now private and Xposed somehow can't hook into it

BlueMax commented:
Your test apk is not signed. 
Signing it with v1/v2/v3 schemes results in 'parse error'. Can you upload a proper one?
Tested on LOS18.1/armv7.
```
W/PackageInstaller(13514): Parse error when parsing manifest. Discontinuing installation
```
Edit: Looks like it doesn't like zipalign even though i did it before apksign (as it's meant to be done for v2+). 
Will test it now...

chaoscalm commented:
> Signal is now the only one acting up, though it seems it switched to using the notification trick to push new message notifications. I can't be bothered to backup my messages and install something else right now.

After a first research, I setted myself with a completely different app that enforce security measures, here is the F-Droid repo:

https://molly.im/fdroid/repo?fingerprint=3B7E93B1FE32C6E35A93D6DDFC5AFBEB1239A7C6EA6AF20FF33ED53CDC38B04A

I think this could help your situation: 

> A workaround for this issue has been reported on r/MicroG and I have verified it to work: 
    https://github.com/microg/GmsCore/issues/1408

    https://www.reddit.com/r/MicroG/comments/kuhgse/device_registration_and_push_notifications/girx53t

whew-inc commented:
@dantecpu Thanks for this. I don't think the microg thing will help because literally all my other apps work now. It's just Signal being Signal I guess.

@BlueMax I've updated the release to include a signed apk. Note that it's signed with debug keys.

BlueMax commented:
Thanks. I've tested it on Android 11 and the self-test is ok now indeed. I don't have any further GCM demanding apps installed atm. 
What i can say though is that it has the same problem as xprivacy with lsposed. You need to explicitly enable all apps that need fake sigs. That's not a module fault of course. 
Thanks for the update.

whew-inc commented:
I will change this pull request to include android 11 support.

Fs00 commented on app/src/main/java/com/thermatk/android/xf/fakegapps/FakeSignatures.java:
```diff
@@ -12,30 +13,39 @@
+
+                            XposedBridge.log("FakeGApps: returning fake signature to \"" + callerAppName + "\" (" + callerPackageName + ")");
                         } catch (NullPointerException e) {
-                            XposedBridge.log("FakeGApps: caught the NullPointerException");
+                            XposedBridge.log("FakeGApps: " + e);
                         }
```
This `catch` block is useless since a NPE can never happen when concatenating strings, and everything else that might throw it is already wrapped in the nested `try`.

Fs00 commented on app/src/main/java/com/thermatk/android/xf/fakegapps/FakeSignatures.java:
```diff
@@ -12,30 +13,39 @@
+                            } catch (Exception ignored) {
+                            }
+
+                            XposedBridge.log("FakeGApps: returning fake signature to \"" + callerAppName + "\" (" + callerPackageName + ")");
```
Could you please log this message to Logcat using `Log.i`? Official docs for `XposedBridge.log` method says:
> **DON'T FLOOD THE LOG!!!** This is only meant for error logging.
> If you want to write information/debug messages, use logcat.

There's also an issue about this (#34), it's very annoying to see all those messages while debugging issues with other modules.

Fs00 commented on app/src/main/java/com/thermatk/android/xf/fakegapps/FakeSignatures.java:
```diff
@@ -12,30 +13,39 @@
+                                callerPackageName = pm.getNameForUid(Binder.getCallingUid());
 
-                            XposedBridge.log("FakeGApps: returning fake signature to " + callerPackageName);
+                                try {
+                                    callerAppName = (String) pm.getApplicationLabel(pm.getApplicationInfo(callerPackageName, 0));
+                                } catch (final PackageManager.NameNotFoundException ignored) {
+                                }
```
This is just a personal opinion - I think that logging the application label does not add any value to the message and makes it unnecessarily wordier when an application label cannot be found (which is not a rare scenario).
Also 99% of the time the package name is clear enough to identify at a glance which app it belongs to.

Fs00 reviewed (commented):
Thanks for your work @whew-inc, I tested your patch on Android 10 and Android 6 and it worked flawlessly.
I've seen also that the ApplicationPackageManager method you're now hooking into is present since Android 4.0, therefore it's super backwards-compatible.
I left some feedback about the code to (IMO) improve this patch even more before it gets merged (hopefully) :D

MrNessuno commented:
> > This is a good change. It wasn't working for me on 11 though. I got it working by modifying the hook to hook ApplicationPackageManger.getPackageInfo
> 
> This seems to work for me too on 11. However I think this won't work on older releases so likely gonna have to add a version check.
> 
> I built an apk for testing: https://github.com/whew-inc/FakeGApps/releases/tag/android11_initial
> 
> Can someone test this with android < 11?
> 
> (If you would rather build it yourself: https://github.com/whew-inc/FakeGApps/tree/android11_initial)

`LSPosed Log
--------- beginning of information
Manufacturer: Xiaomi
Brand: Redmi
Device: lmi
Product: lmi
Model: Redmi K30 Pro
Fingerprint: Redmi/lmi/lmi:11/RKQ1.200826.002/V12.2.3.0.RJKCNXM:user/release-keys
ROM description: lmi-user 11 RKQ1.200826.002 V12.2.3.0.RJKCNXM release-keys
Architecture: arm64-v8a
Android build: RKQ1.200826.002
Android version: 11
Android sdk: 30
LSPosed version: v1.3.4 (5501)
Riru version: v25.3.4.r399.84f7084 (399)
Riru api: 25
Magisk: 22.0 (22000)
2021-04-03 19:34:01.001 I/?(1968-1968): FakeGApps: returning fake signature to "unknown package label" (android.uid.system:1000)`

Is EdExposed better than LSPosed? I tryed with the 2.0 and the new test apk but with this new one I got this error, what can I try?

whew-inc commented:
@MrNessuno It's normal. This happens when the sig is spoofed for system itself. Check if eg. Microg works. @Fs00' feedback fixes this but I haven't found the time to get onto it yet.

MrNessuno commented:
> @MrNessuno It's normal. This happens when the sig is spoofed for system itself. Check if eg. Microg works. @Fs00' feedback fixes this but I haven't found the time to get onto it yet.

I have to enable the module and check System Framework right? I tried with the Signature Spoofing Checker and I got disabled as result

whew-inc commented:
> > @MrNessuno It's normal. This happens when the sig is spoofed for system itself. Check if eg. Microg works. @Fs00' feedback fixes this but I haven't found the time to get onto it yet.
> 
> I have to enable the module and check System Framework right? I tried with the Signature Spoofing Checker and I got disabled as result

I don't know how Lxposed works but in Edxposed all you have to do is enable the module and reboot.

This module only spoofs two google play apps (play store and services). Because Signature Spoofing Checker isn't one of the two, its signature won't be spoofed and therefore the app thinks spoofing doesn't work.

chaoscalm commented:
> > > @MrNessuno It's normal. This happens when the sig is spoofed for system itself. Check if eg. Microg works. @Fs00' feedback fixes this but I haven't found the time to get onto it yet.
> > 
> > 
> > I have to enable the module and check System Framework right? I tried with the Signature Spoofing Checker and I got disabled as result
> 
> I don't know how Lxposed works but in Edxposed all you have to do is enable the module and reboot.
> 
> This module only spoofs two google play apps (play store and services). Because Signature Spoofing Checker isn't one of the two, its signature won't be spoofed and therefore the app thinks spoofing doesn't work.

LSPosed is the opposite approach to EdXposed, by default nothing is hooked, you have to select it, so from a security/bug point of view is better, I remember on EdXposed a similar options with it working fine, also we can make rules for it for automatic selection of best settings.

**Edit:** Indeed the verbose log isn't something new I think? Anyway in order for LSPosed to work with fakegapps, I added what is required to be enabled on Android 11.

**Edit:** 

> > > This is a good change. It wasn't working for me on 11 though. I got it working by modifying the hook to hook ApplicationPackageManger.getPackageInfo
> > 
> > 
> > This seems to work for me too on 11. However I think this won't work on older releases so likely gonna have to add a version check.
> > I built an apk for testing: https://github.com/whew-inc/FakeGApps/releases/tag/android11_initial
> > Can someone test this with android < 11?
> > (If you would rather build it yourself: https://github.com/whew-inc/FakeGApps/tree/android11_initial)
> 
> `LSPosed Log --------- beginning of information Manufacturer: Xiaomi Brand: Redmi Device: lmi Product: lmi Model: Redmi K30 Pro Fingerprint: Redmi/lmi/lmi:11/RKQ1.200826.002/V12.2.3.0.RJKCNXM:user/release-keys ROM description: lmi-user 11 RKQ1.200826.002 V12.2.3.0.RJKCNXM release-keys Architecture: arm64-v8a Android build: RKQ1.200826.002 Android version: 11 Android sdk: 30 LSPosed version: v1.3.4 (5501) Riru version: v25.3.4.r399.84f7084 (399) Riru api: 25 Magisk: 22.0 (22000) 2021-04-03 19:34:01.001 I/?(1968-1968): FakeGApps: returning fake signature to "unknown package label" (android.uid.system:1000)`
> 
> Is EdExposed better than LSPosed? I tryed with the 2.0 and the new test apk but with this new one I got this error, what can I try?

If the log trouble you, you can try merge this into 11 and test, https://github.com/eebssk1/FakeGApps/commit/3b8aebc0efd39c968905ca62b430ba182f5af95e

**Final Result:** both the log problem you mentioned, and the LSPosed are functional with the Android 11 branch.

![photo_2021-04-04_02-50-27](https://user-images.githubusercontent.com/7214961/113495298-f3e70e80-94f0-11eb-8e33-0c9789793995.jpg)
![photo_2021-04-04_02-55-05](https://user-images.githubusercontent.com/7214961/113495329-3f99b800-94f1-11eb-832b-9fcbbabaeb34.jpg)

MrNessuno commented:
where is recommended to download MicroG to be more compatible with this module? (sorry for "spamming" here, but idk another place for this question)

chaoscalm commented:
> where is recommended to download MicroG to be more compatible with this module? (sorry for "spamming" here, but idk another place for this question)

I personally use this: https://github.com/FriendlyNeighborhoodShane/MinMicroG-abuse-CI

You are welcome.

chaoscalm commented:
> > where is recommended to download MicroG to be more compatible with this module? (sorry for "spamming" here, but idk another place for this question)
> 
> I personally use this: https://github.com/FriendlyNeighborhoodShane/MinMicroG-abuse-CI
> 
> You are welcome.

ps: nobody recommended me anything, I just tested them during time, in the end I developed a personal taste for the one I shared with you, which of course maybe not of your taste, another one is nanodroid...

MrNessuno commented:
> > > where is recommended to download MicroG to be more compatible with this module? (sorry for "spamming" here, but idk another place for this question)
> > 
> > 
> > I personally use this: https://github.com/FriendlyNeighborhoodShane/MinMicroG-abuse-CI
> > You are welcome.
> 
> ps: nobody recommended me anything, I just tested them during time, in the end I developed a personal taste for the one I shared with you, which of course maybe not of your taste, another one is nanodroid...

Is it true that I have to rename the zip file with system-minmicrog ecc ecc.zip before flashing? I saw it in XDA but I don't see that thing in github

chaoscalm commented:
> > > > where is recommended to download MicroG to be more compatible with this module? (sorry for "spamming" here, but idk another place for this question)
> > > 
> > > 
> > > I personally use this: https://github.com/FriendlyNeighborhoodShane/MinMicroG-abuse-CI
> > > You are welcome.
> > 
> > 
> > ps: nobody recommended me anything, I just tested them during time, in the end I developed a personal taste for the one I shared with you, which of course maybe not of your taste, another one is nanodroid...
> 
> Is it true that I have to rename the zip file with system-minmicrog ecc ecc.zip before flashing? I saw it in XDA but I don't see that thing in github

I don't know, personally I never rename anything, had builded what I shared with you too inside VM and lately using the one provided with the github-actions, if you can share the XDA post probably I can tell you, but based on my experience no.

MrNessuno commented:
I installed MicroG removing Google Apps and... It works! :cupid: and with MIUI is the only patch that works ^^

chaoscalm commented:
> I installed MicroG removing Google Apps and... It works! 💘 and with MIUI is the only patch that works ^^

I am happy for your success Mr. Nessuno, sometime we get lucky.

MrNessuno commented:
How to type *#*#CHECKIN#*#* in the dialer?? if I copy paste it I only get this *#*#N#*#* and I see that apps aren't showing in Cloud Messaging in MicroG and I receive notifications only when apps are still in RAM

whew-inc committed:
Remove redundant NPE try-catch block

whew-inc committed:
Remove application label logging

whew-inc committed:
Use Log.i instead of XposedBridge.log

whew-inc committed:
Add LSposed module scope

whew-inc commented:
No log spam and LSposed scoped module support: https://github.com/whew-inc/FakeGApps/releases/tag/lsposed

Fs00 commented:
Awesome job @whew-inc! Thanks for listening to my feedback 😃

@thermatk Could you please look into this PR and merge it if it's fine for you?

Fs00 committed:
Hook PackageManagerService class in system server

This avoids the need to select every app that needs GMS signature
spoofing when using LSPosed

Fs00 committed:
Add Android Studio project folder to gitignore

Fs00 committed:
Bump version

BlueDrink9 commented:
I have installed this fork, but I still don't get signature spoofing. When I enter the code in the dialer, nothing happens. Is there another way to trigger it other than the dialer code? Or does this sound like a different issue?

whew-inc committed:
Upgrade Gradle to 7.0.2

whew-inc commented:
https://github.com/whew-inc/FakeGApps/pull/1#issuecomment-940050877

Looking for someone to test this with **LSposed** on **Android 11**.

Improved app startup performance is one of the changes

whew-inc committed:
Merge pull request #1 from Fs00/lsposed

Change hook logic to avoid having to enable the module for all apps on LSPosed

whew-inc commented:
New release: https://github.com/whew-inc/FakeGApps/releases/tag/3.0

kurt-by commented:
Hi @whew-inc 
Your log fix release works fine on android 12. The 3.0 version doesn't. No mattee what i select in apps list.

nift4 committed:
Update gradle plugin to 7.1.3

nift4 committed:
Fixed all gradle warnings; added signing config

nift4 committed:
Port to Android 12

nift4 committed:
Bump version to 3.1

nift4 committed:
Update README.md

nift4 committed:
Remove logspam, bump version

nift4 committed:
Rework to unbreak Android 11 (needs testing)

whew-inc committed:
Class hook depends on Android version running on device

whew-inc committed:
Bump version

whew-inc committed:
Hook PackageManagerService on all versions below Android S

whew-inc committed:
Merge pull request #4 from nift4/master

Port to Android 12

whew-inc committed:
Merge pull request #5 from whew-inc/compatibility

Compatibility

whew-inc committed:
Package namespace change

Preparation to submit the module to LSPosed repo

whew-inc committed:
Update README.md

whew-inc committed:
Upgrade Gradle to 7.2.1

whew-inc committed:
Bump module version to 4.1

whew-inc commented:
@kurt-by The module has been updated for Android 12.

whew-inc commented:
Closing as there has been no response from the author for 2+ years now.
